Improve ":id" param normalization

soyuka opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes/no (depends)
| BC breaks?    | yes
| Deprecations? | yes
| Tests pass?   | not yet
| Fixed tickets | #1309 #1476 and maybe other (#1376 start) #1234 #1309 
| License       | MIT
| Doc PR        | n/a

## Issue

Currently, as I mentioned in #1309, our `DataProviders` get the raw `id` param coming from the request. ie https://github.com/api-platform/core/blob/183526946637ecf72ce1ddc4032ede281ee3b381/src/EventListener/ReadListener.php#L103

There is an issue here where we sometimes want our `DataProviders` to get a correct instance of the given identifier (eg `Ramsey\UUID` for an uuid or `\DateTime` for a date).

## What's being done

This is what I've come up with. Basically it adds a `IdentifierNormalizer` class (does not implement the `NormalizerInterface` because it's not one) that takes the `string` from `$request->attributes->get('id')` and returns an `array` with correct Types. 
This is not working yet because I have a train to catch haha. Going to add tests and the ability to add normalizers asap (those are `NormalizerInterface`, for example `DateTimeNormalizer` etc.)!

There is a BC break but I've added deprecation and my goal is to keep everything working as before by getting rid of the `IdentifierManagerTrait` in future versions.

## Benefits

With this fix the identifier normalization isn't linked to Doctrine anymore. This is a great improvement for future `DataProviders` that might need the correct types without forcing the developer to use Doctrine. Also, it allows the developper to add custom normalizers for identifiers (for example the UUID mentioned above). 
By having a cleaner way to parse those, it'll also help us manage errors and throw a `NotFoundException` when needed (mentioned by @theofidry in #1309).
Last but not least, I've improved the composite identifier parser as for now something like `a=test;b=bar ;foo;c=123;459;` would totally fail because of the `;` in the identifier value (nb try with `=`).

ping @dkarlovi

mickaelandrieu commented on src/Identifier/CompositeIdentifierNormalizer.php:
```diff
@@ -0,0 +1,44 @@
+    }
+}
+
+// $id = new IdentifierNormalizer();
```
to be removed, isn't it? :)

soyuka commented:
@mickaelandrieu haha of course sorry!

PR updated with tests, a compiler pass, correct deprecation, BC break coverage

I just need to update the tests for the deprecation and open a doc PR!

Let me know your thoughts @api-platform/core-team !

soyuka commented on src/Bridge/Symfony/Bundle/Resources/config/api.xml:
```diff
@@ -218,6 +219,19 @@
+
+        <service id="api_platform.identifier.date_normalizer" class="Symfony\Component\Serializer\Normalizer\DateTimeNormalizer" public="false">
+            <tag name="api_platform.identifier_normalizer" />
+        </service>
```
I'm not sure that this is allowed, lmk if there's something better to use instead.

norkunas replied:
What about adding uuid normalizer? It is very common I think.

soyuka replied:
added below :) Just need to test it!

dunglas replied:
Why would not it be?

dkarlovi commented:
This looks great @soyuka!

About the "Not found" when un-normalizable (that's actually a word, or my Chrome spell-checker is asleep), maybe introducing a new Exception will enable the desired behaviour, it can be hooked up to a 404 status by default.

soyuka commented:
TODO:
- [ ] Test uuid denormalizer
- [ ] Improve deprecations

soyuka commented on src/Bridge/Doctrine/Orm/ItemDataProvider.php:
```diff
@@ -65,11 +65,18 @@ public function supports(string $resourceClass, string $operationName = null): b
      * @throws RuntimeException
      */
-    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = [])
+    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = []/** , array $identifiers **/)
```
This was the only solution I had to avoid breaking the `DataProvider::getItem` signature. Also, in future versions `$identifier` should replace `$id` and it will be typed as `array`.

norkunas commented:
Yes, I'm using it already;)

soyuka commented:
@norkunas could you try this branch to see if it fixes your issue? I think that my implementation is as far as it will get :).

norkunas commented:
I'll try this evening.

dkarlovi commented:
@soyuka I'm testing it now.

dkarlovi commented:
It works for me (as in, string gets converted to a valid `Ramsey\Uuid\Uuid` instance), but the lookup fails because of https://github.com/ramsey/uuid-doctrine/issues/18. Would probably need to create a custom normalizer which isn't that bad.

LGTM!

soyuka commented:
Not sure I understood the issue @dkarlovi is it something that needs to be done here? Lmk if needed and if you have a failing example.

dkarlovi commented:
I think it's Doctrine's fault, it's not able to create the proper value to do a lookup with so you need to supply it yourself.

To be exact, I'm using [binary columns](https://github.com/ramsey/uuid-doctrine#binary-database-columns) to represent IDs (so, they're not in their string representation in the DB). Doctrine should be able to serialize the object properly before querying, but it's not.

I don't think you should change anything here.

soyuka commented:
@dkarlovi IMO this should be done in the custom `UuidType` where `convertToDatabaseValue` should call `getBytes`.

dkarlovi commented:
I think Doctrine should fix the underlying issue as there are plenty of places that get bitten by this, but your proposed solution works in the meantime.

In any case, that's outside of scope here: your API Platform fix works as expected and IMHO it can be merged as-is.Thank you for doing it!  :+1:

dunglas commented on src/Identifier/CompositeIdentifierNormalizer.php:
```diff
@@ -0,0 +1,39 @@
+/**
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+class CompositeIdentifierNormalizer
```
`final`

dunglas commented on src/Identifier/IdentifierNormalizer.php:
```diff
@@ -0,0 +1,90 @@
+/**
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+class IdentifierNormalizer implements IdentifierNormalizerInterface
```
`final`

dunglas commented on src/Exception/InvalidIdentifierException.php:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+class InvalidIdentifierException extends \Exception implements ExceptionInterface
```
`final`

dunglas commented on src/Bridge/Ramsey/Normalizer/UuidNormalizer.php:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+class UuidNormalizer implements DenormalizerInterface
```
`final`

dunglas commented on src/Identifier/IdentifierNormalizer.php:
```diff
@@ -0,0 +1,90 @@
+{
+    private $propertyNameCollectionFactory;
+    private $propertyMetadataFactory;
+    private $compositeIdentifierNormalizer;
```
This class has no interface, maybe can it be initialized directly in the constructor instead of being injected?

soyuka commented on src/Bridge/Doctrine/Orm/Util/IdentifierManagerTrait.php:
```diff
@@ -39,6 +39,8 @@
      */
     private function normalizeIdentifiers($id, ObjectManager $manager, string $resourceClass): array
     {
+        @trigger_error(sprintf('Trait ApiPlatform\Core\Bridge\Doctrine\Orm\Util\IdentifierManagerTrait will be removed in version 3.0. Using it is deprecated since 2.2.'), E_USER_DEPRECATED);
```
Nahh I can't as "php-cs-fixer" is using `self` instead but because it's a trait the resulting class for `self::class` is wrong :|. IMO it's a bug in `php-cs-fixer`.

dunglas replied:
Can't you just disable this rule for now?

backbone87 commented:
Sorry that i swing in so late into this PRs progress, but I think the interface should get the whole `Request` object.

Why is that? It would allow one to write a normalizer that can handle stuff like this: `/parent/{parent_id}/child/{child_id}` where childs identifier is composite.

Also for some more non-standard use-cases like taking identifier information from the query string.

dunglas commented:
@backbone87 you can use the `request_stack` service for that.

soyuka commented:
@backbone87 yes I was thinking about this use case as well while working on this. However, I'm not sure what `interface` you're talking about. You can overwrite/inject `request_stack` if needed but I doubt it'll be sufficient :|.

If I take your example `/parent/{parent_id}/child/{child_id}` there are a few things that can/should be handled before it can work.

First, today we still rely on [`$id = $request->attributes->get('id');`](https://github.com/soyuka/core/blob/2751a870f2b7ddaa41c9e20ab983edf8c813e7a5/src/EventListener/ReadListener.php#L111) to get the "identifier".
Then, the resource url is created through the `operationPathResolver` (see ApiLoader).

I see an option that could work with this PR, having a custom operation, with `:id` of format `{parent_id}/child/{child_id}` (not sure if `/` can be part of a parameter haha). Then one would need to add an `IdentifierNormalizer` that parses this format.

backbone87 commented:
> today we still rely on $id = $request->attributes->get('id'); to get the "identifier".

this is pretty simple to solve. with this PR the resource gets an already normalized id array. for all the other places that uses `$request->attributes->get('id')` we can provide a fallback:
```php
$identifiers = $normalizer->normalize($request, $resouceClass);
$request->attributes->set('id', http_build_query($identifiers, '', ';'));
```

> then, the resource url is created through the operationPathResolver (see ApiLoader).

this maybe could be solved by providing a "partial" denormalize, which maps the composite ID's keys to the route params without actually doing the `http_build_query` (for id route param). that "partially" denormalized array can be passed to the url generator

backbone87 commented:
> today we still rely on $id = $request->attributes->get('id'); to get the "identifier".

this is pretty simple to solve. with this PR the resource gets an already normalized id array. for all the other places that uses `$request->attributes->get('id')` we can provide a fallback:
```php
$identifiers = $normalizer->normalize($request, $resouceClass);
$request->attributes->set('id', http_build_query($identifiers, '', ';'));
```

> then, the resource url is created through the operationPathResolver (see ApiLoader).

this maybe could be solved by providing a "partial" denormalize, which maps the composite ID's keys to the route params without actually doing the `http_build_query` (for id route param). that "partially" denormalized array can be passed to the url generator

backbone87 commented:
btw. i dont want endpoints like `/parent/{pid}/child/{cid}` to be automatically resolved, but explicitly given in the operation config

soyuka commented:
Maybe we could just use `$identifiers = $this->identifierNormalizer->normalize($request, $attributes['resource_class']);` this would indeed give much more power to the end user.
The default normalizer would still use `$request->attributes->get('id')` but you'll be able to overwrite the normalizer to do what you want.

WDYT?

soyuka commented:
ping @api-platform/core-team would love some reviews on this :)

dunglas commented on composer.json:
```diff
@@ -64,7 +64,9 @@
         "symfony/yaml": "^3.3 || ^4.0",
-        "webonyx/graphql-php": "^0.10.2"
+        "webonyx/graphql-php": "^0.10.2",
+        "ramsey/uuid": "^3.7",
```
Can you order the deps alphabetically please?

dunglas replied:
You should also add it in the suggest section.

dunglas commented on composer.json:
```diff
@@ -79,7 +81,8 @@
         "symfony/twig-bundle": "To use the Swagger UI integration.",
-        "webonyx/graphql-php": "To support GraphQL."
+        "webonyx/graphql-php": "To support GraphQL.",
+        "ramsey/uuid": "To support uuid identifiers."
```
Same here.

dunglas commented on src/Bridge/Doctrine/Orm/ItemDataProvider.php:
```diff
@@ -66,11 +66,18 @@ public function supports(string $resourceClass, string $operationName = null): b
      * @throws RuntimeException
      */
-    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = [])
+    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = []/**, array $identifiers = []**/)
```
Can't you pass the identifiers in the context instead?

dunglas commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -106,7 +106,13 @@ public function getSubresource(string $resourceClass, array $identifiers, array
+
+            if (isset($identifiers[$identifier])) {
+                // if it's an array it's already normalized, the IdentifierManagerTrait is deprecated
+                $normalizedIdentifiers = is_array($identifiers[$identifier]) ? $identifiers[$identifier] : $this->normalizeIdentifiers($identifiers[$identifier], $manager, $identifierResourceClass);
```
`\is_array` (opcache optim).

dunglas commented on src/Bridge/Doctrine/Orm/Util/IdentifierManagerTrait.php:
```diff
@@ -39,6 +39,8 @@
      */
     private function normalizeIdentifiers($id, ObjectManager $manager, string $resourceClass): array
     {
+        @trigger_error(sprintf('Trait "%s" will be removed in version 3.0. Using it is deprecated since 2.2.', IdentifierManagerTrait::class), E_USER_DEPRECATED);
```
This trait is internal, so there is no need to deprecate it, you can even remove it if you want to.

soyuka replied:
If I remove the trait, injecting the IdentifierNormalizer will be mandatory

dunglas replied:
All the dependencies of the composite identifier normalizer are already injected in `SubresourceDataProvider`, you can just initialize the class in the constructor if `null` is passed (and trigger a deprecation notice). The other benefit of doing this is that we'll have only one implementation to maintain.

dunglas replied:
What about this one @soyuka?

dunglas commented on src/Bridge/Ramsey/Normalizer/UuidNormalizer.php:
```diff
@@ -0,0 +1,42 @@
+     */
+    public function supportsDenormalization($data, $type, $format = null)
+    {
+        return interface_exists(UuidInterface::class) && (UuidInterface::class === $type || is_subclass_of($type, UuidInterface::class));
```
I think you can remove the `interface_exists` call (at least, you should check `UuidInterface::class === $type` first because it's less coastly).

dunglas commented on src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/IdentifierNormalizerPass.php:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Injects identifier normalizers.
+ *
+ * @internal
```
Can't it be part of the public API?

dunglas commented on src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/IdentifierNormalizerPass.php:
```diff
@@ -0,0 +1,40 @@
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class IdentifierNormalizerPass implements CompilerPassInterface
```
Maybe can you bump the dep `symfony/di` to 3.4 and use https://symfony.com/blog/new-in-symfony-3-4-simpler-injection-of-tagged-services instead.

soyuka replied:
Why don't we keep this for another PR where we use this improvement on every cases we have?

dunglas replied:
If we introduce a new class, we need to maintain it (and deprecate it soon), but your call.

dunglas commented on src/Bridge/Symfony/Routing/IriConverter.php:
```diff
@@ -73,7 +76,13 @@ public function getItemFromIri(string $iri, array $context = [])
-        if ($item = $this->itemDataProvider->getItem($parameters['_api_resource_class'], $parameters['id'], null, $context)) {
+        $identifiers = [];
+
+        if ($this->identifierNormalizer) {
```
Can't we do this directly in the data provider? It would encapsulate the logic in only one place.

dunglas commented on src/DataProvider/ChainItemDataProvider.php:
```diff
@@ -35,16 +35,21 @@ public function __construct(array $dataProviders)
      * {@inheritdoc}
      */
-    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = [])
+    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = []/**, array $identifiers = [] **/)
```
IMO we should just normalize the identifiers directly in the `ItemDataProvider` (or use the context), but I would like to get rid of this extra parameter.

dunglas commented on src/EventListener/ReadListener.php:
```diff
@@ -33,12 +34,19 @@
         $this->itemDataProvider = $itemDataProvider;
         $this->subresourceDataProvider = $subresourceDataProvider;
+
+        if (null === $identifierNormalizer) {
```
Is this mandatory, can't we just do nothing if `null` (in many cases, you don't need to normalize the identifier at all).

dunglas commented on src/EventListener/ReadListener.php:
```diff
@@ -133,10 +146,18 @@ private function getSubresourceData(Request $request, array $attributes)
-        foreach ($attributes['subresource_context']['identifiers'] as $key => list($id, , $hasIdentifier)) {
-            if (true === $hasIdentifier) {
-                $identifiers[$id] = $request->attributes->get($id);
+        try {
```
Can't we have a less broad try?

dunglas commented on src/EventListener/ReadListener.php:
```diff
@@ -133,10 +146,18 @@ private function getSubresourceData(Request $request, array $attributes)
-            if (true === $hasIdentifier) {
-                $identifiers[$id] = $request->attributes->get($id);
+        try {
+            foreach ($attributes['subresource_context']['identifiers'] as $key => list($id, $resourceClass, $hasIdentifier)) {
```
```php
if (!$hasIdentifier) {
    continue;
}

$identifiers[$id] = $request->attributes->get($id);
if ($this->identifierNormalizer) {
    $identifiers[$id] = $this->identifierNormalizer->normalize((string) $identifiers[$id], $resourceClass);
}
```
$identifiers[$id] = $this->identifierNormalizer ?

dunglas commented on src/Identifier/IdentifierNormalizer.php:
```diff
@@ -0,0 +1,92 @@
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Identiifer normalizer.
```
typo (I suggest: `Normalizes an indentifier`)

dunglas commented on src/Identifier/IdentifierNormalizer.php:
```diff
@@ -0,0 +1,92 @@
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class IdentifierNormalizer implements IdentifierNormalizerInterface
```
`ChainIdentifierNormalizer` for consistency?

dunglas commented on src/Identifier/IdentifierNormalizer.php:
```diff
@@ -0,0 +1,92 @@
+            }
+
+            $type = $propertyMetadata->getType();
+            $metadataIdentifiers[$propertyName] = !$type ?: (Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType() ? $type->getClassName() : null);
```
Can you switch to if statements (nested ternaries are very hard to read and throw warning in many linters).

dunglas commented on src/Identifier/IdentifierNormalizer.php:
```diff
@@ -0,0 +1,92 @@
+                if ($normalizer->supportsDenormalization($identifier, $metadataIdentifiers[$key])) {
+                    try {
+                        $identifiers[$key] = $normalizer->denormalize($identifier, $metadataIdentifiers[$key]);
+                    } catch (\Exception $e) {
```
When can this occur? Catching `\Exception` is usually broader and is a pain to debug.

soyuka replied:
Agreed. In that case I'd have to improve the interface and we have to throw an `InvalidIdentifierException` in the denormalizer (fix #1476)

dunglas commented on src/Bridge/Symfony/Bundle/Resources/config/api.xml:
```diff
@@ -220,6 +222,22 @@
+            <tag name="api_platform.identifier_normalizer" />
+        </service>
+
+        <service id="api_platform.identifier.uuid_normalizer" class="ApiPlatform\Core\Bridge\Ramsey\Normalizer\UuidNormalizer" public="false">
```
You must remove this service in the extension if Ramsey's UUID isn't installed.

soyuka commented:
Thanks for the review @dunglas ! I changed the detection in the dataproviders by using the context! 

I don't think it's the data provider responsability to normalize identifiers. We should get the correct identifiers in the data providers as it'd simplify DX when using our own data providers !

TODO: review identifier normalizer exceptions when denormalizing (remove `\Exception`).

soyuka commented on src/EventListener/ReadListener.php:
```diff
@@ -101,10 +110,17 @@ private function getCollectionData(Request $request, array $attributes)
+            }
+
+            $data = $this->itemDataProvider->getItem($attributes['resource_class'], $identifiers ?: $id, $attributes['item_operation_name'], $context);
+        } catch (InvalidIdentifierException $e) {
```
@dunglas this is how I'd like things to be, although I think that this will break current DataProvider implementations. 
If we keep this in the data provider, implementing gets more complex and this path improves DX (end developper doesn't care about identifiers and gets what he expects, for example Uuid/Datetime instance instead of string).

soyuka replied:
ping @dunglas thoughts about using `$context[ChainIdentifierNormalizer::HAS_IDENTIFIER_NORMALIZER]`?

dunglas replied:
Looks good to me.

Gounlaf commented:
Hi,

I've question about this PR - it's not helping for review, sorry =/
I've read the What's being done/Benefits parts, but I still don't get it clearly.

Does this PR will allow us to "identify" an entity with another field than the id/PK?
**Example**: having "regular" ids (integer) as PK, and an UUID field, this one used for identifying the entity from "outside" (explained better in api-platform/api-platform#462).

Thanks

soyuka commented:
@Gounlaf not really, this patch is only there to improve the "identifier" normalization (ie: ':id' part of `/api/foo/:id`). In api-platform/api-platform#462 the uuid field is the real identifier.

dunglas commented on .php_cs.dist:
```diff
@@ -55,6 +55,7 @@ return PhpCsFixer\Config::create()
         'strict_comparison' => true,
         'strict_param' => true,
         'ternary_to_null_coalescing' => true,
+        'self_accessor' => false
```
Why this change, the `+` is a bug right?

dunglas commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,99 @@
+    private $compositeIdentifierNormalizer;
+    private $identifierNormalizers;
+
+    public const HAS_IDENTIFIER_NORMALIZER = 'has_normalized_identifier';
```
Shouldn't this const be defined directly in the interface?

dunglas commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,99 @@
+    {
+        $this->propertyNameCollectionFactory = $propertyNameCollectionFactory;
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+        $this->compositeIdentifierNormalizer = new CompositeIdentifierNormalizer();
```
Why not injecting it in the existing array?

soyuka replied:
Because it's not implementing the IdentifierNormalizerInterface but it's called pre-normalization when we have more then 1 identifier. See  https://github.com/api-platform/core/pull/1478/files/b3037d9fa3ec663a8f1dd51c771e67a3aeab23f5#diff-9622bccedd195e39c3b3da920161ef42R69

dunglas replied:
Why does it have a different behavior than other normalizers? It looks weird to me.

soyuka commented:
ping @dkarlovi

dkarlovi commented:
Following up the discussion from Twitter yesterday:

I've updated to 2.1.4 and now getting:

```
Ramsey\Uuid\Exception\InvalidUuidStringException: Invalid UUID string: ,urdpfiń��˗�� in /app/vendor/ramsey/uuid/src/Codec/StringCodec.php:143
Stack trace:
#0 /app/vendor/ramsey/uuid/src/Codec/StringCodec.php(80): Ramsey\Uuid\Codec\StringCodec->extractComponents(',urdpfi\xC5\x84\x9B\x15\x8B\xCB\x97\xD2...')
#1 /app/vendor/ramsey/uuid/src/UuidFactory.php(191): Ramsey\Uuid\Codec\StringCodec->decode(',urdpfi\xC5\x84\x9B\x15\x8B\xCB\x97\xD2...')
#2 /app/vendor/ramsey/uuid/src/Uuid.php(629): Ramsey\Uuid\UuidFactory->fromString(',urdpfi\xC5\x84\x9B\x15\x8B\xCB\x97\xD2...')
#3 /app/vendor/ramsey/uuid-doctrine/src/UuidBinaryType.php(94): Ramsey\Uuid\Uuid::fromString(',uRdpFI\xC5\x84\x9B\x15\x8B\xCB\x97\xD2...')
#4 /app/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(1574): Ramsey\Uuid\Doctrine\UuidBinaryType->convertToDatabaseValue(',uRdpFI\xC5\x84\x9B\x15\x8B\xCB\x97\xD2...', Object(Doctrine\DBAL\Platforms\MySQL57Platform))
#5 /app/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(1539): Doctrine\DBAL\Connection->getBindingInfo(',uRdpFI\xC5\x84\x9B\x15\x8B\xCB\x97\xD2...', Object(Ramsey\Uuid\Doctrine\UuidBinaryType))
#6 /app/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(906): Doctrine\DBAL\Connection->_bindTypedValues(Object(Doctrine\DBAL\Driver\PDOStatement), Array, Array)
#7 /app/vendor/doctrine/orm/lib/Doctrine/ORM/Query/Exec/SingleSelectExecutor.php(50): Doctrine\DBAL\Connection->executeQuery('SELECT i0_.name...', Array, Array, NULL)
#8 /app/vendor/doctrine/orm/lib/Doctrine/ORM/Query.php(321): Doctrine\ORM\Query\Exec\SingleSelectExecutor->execute(Object(Doctrine\DBAL\Connection), Array, Array)
#9 /app/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php(969): Doctrine\ORM\Query->_doExecute()
#10 /app/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php(924): Doctrine\ORM\AbstractQuery->executeIgnoreQueryCache(NULL, NULL)
#11 /app/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php(766): Doctrine\ORM\AbstractQuery->execute(NULL, NULL)
#12 /app/vendor/api-platform/core/src/Bridge/Doctrine/Orm/ItemDataProvider.php(97): Doctrine\ORM\AbstractQuery->getOneOrNullResult()
#13 /app/vendor/api-platform/core/src/DataProvider/ChainItemDataProvider.php(42): ApiPlatform\Core\Bridge\Doctrine\Orm\ItemDataProvider->getItem('ImageBundle\\Ent...', '2c755264-7046-4...', 'delete', Array)
#14 /app/vendor/api-platform/core/src/EventListener/ReadListener.php(106): ApiPlatform\Core\DataProvider\ChainItemDataProvider->getItem('ImageBundle\\Ent...', '2c755264-7046-4...', 'delete')
#15 /app/vendor/api-platform/core/src/EventListener/ReadListener.php(64): ApiPlatform\Core\EventListener\ReadListener->getItemData(Object(Symfony\Component\HttpFoundation\Request), Array)
#16 [internal function]: ApiPlatform\Core\EventListener\ReadListener->onKernelRequest(Object(Symfony\Component\HttpKernel\Event\GetResponseEvent), 'kernel.request', Object(Symfony\Component\HttpKernel\Debug\TraceableEventDispatcher))
#17 /app/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/WrappedListener.php(104): call_user_func(Array, Object(Symfony\Component\HttpKernel\Event\GetResponseEvent), 'kernel.request', Object(Symfony\Component\HttpKernel\Debug\TraceableEventDispatcher))
#18 [internal function]: Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke(Object(Symfony\Component\HttpKernel\Event\GetResponseEvent), 'kernel.request', Object(Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher))
#19 /app/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php(212): call_user_func(Object(Symfony\Component\EventDispatcher\Debug\WrappedListener), Object(Symfony\Component\HttpKernel\Event\GetResponseEvent), 'kernel.request', Object(Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher))
#20 /app/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php(44): Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(Array, 'kernel.request', Object(Symfony\Component\HttpKernel\Event\GetResponseEvent))
#21 /app/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php(139): Symfony\Component\EventDispatcher\EventDispatcher->dispatch('kernel.request', Object(Symfony\Component\HttpKernel\Event\GetResponseEvent))
#22 /app/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php(129): Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch('kernel.request', Object(Symfony\Component\HttpKernel\Event\GetResponseEvent))
#23 /app/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php(68): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#24 /app/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php(169): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#25 /app/web/index.php(26): Symfony\Component\HttpKernel\Kernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#26 {main}
```

What I did on 2.1.3 was override `ApiPlatform\Core\Bridge\Doctrine\Orm\Util\IdentifierManagerTrait` with exactly the same copy, only changed

```php
$identifier = DBALType::getType($doctrineTypeName)->convertToPHPValue($identifier, $platform);
```

to

```php
$identifier = DBALType::getType($doctrineTypeName)->convertToDatabaseValue($identifier, $platform);
```

This hack no longer works on 2.1.4, nor does it work without it. So, my question on Twitter was: what changed in 2.1.4 to re-break this?

About your patch, I've tested it before and it works for me, so you want me to test it again?

soyuka commented:
> About your patch, I've tested it before and it works for me, so you want me to test it again?

Nope that's fine I know that this patch is a better replacement for the trait you have issues with.

I'm not sure I get how you modified the trait :P. 

> This hack no longer works on 2.1.4, nor does it work without it. So, my question on Twitter was: what changed in 2.1.4 to re-break this?

Might be https://github.com/api-platform/core/commit/ed4c6887c9be20945f0c012772a7b7de42032020 but it's better now then it was before.

Anyway, for identifiers that are no string/integer I think that this patch is a must have. Though I won't be able to make it pass in `2.1` :|. Would you be able to jump on 2.2 beta?

dkarlovi commented:
> I'm not sure I get how you modified the trait :P.

Copied to my codebase, pointed composer to it, excluded API Platform's copy. :|

> Might be ed4c688 but it's better now then it was before.

Thanks, I'll have a look.

>  Would you be able to jump on 2.2 beta?

Probably yes, will be having a round of upgrades to Symfony 3.4 / 4.0 soon anyway.

soyuka commented:
> Copied to my codebase, pointed composer to it, excluded API Platform's copy. :|

Doh, would probably be better to use a fork directly.

dkarlovi commented:
Will this get merged before 2.2.0 final?

soyuka commented:
> Will this get merged before 2.2.0 final?

I'm not sure, I think @dunglas froze the features releases for now :/. I'm maintaining my fork with it for the time being.

bendavies commented:
What's left to do here?

dkarlovi commented:
Wait. :(

bendavies commented:
Yes, my question is for for? 😁

soyuka commented:
> Yes, my question is what for?

Reviews? :D

dunglas commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -152,11 +153,17 @@ private function buildQuery(array $identifiers, array $context, QueryNameGenerat
+
+        if (isset($identifiers[$identifier])) {
+            // if it's an array it's already normalized, the IdentifierManagerTrait is deprecated
+            $normalizedIdentifiers = ($context[IdentifierNormalizerInterface::HAS_IDENTIFIER_NORMALIZER] ?? false) ? $identifiers[$identifier] : $this->normalizeIdentifiers(
```
Can you put everything on a single line?

dunglas commented on src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/IdentifierNormalizerPass.php:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class IdentifierNormalizerPass implements CompilerPassInterface
```
Can you also make this class `@internal`? When we'll bump the DI component to 4.1, we'll be able to use http://symfony.com/doc/3.4/service_container/tags.html#reference-tagged-services.
Alternatively, we can bump the version now and remove this class.

bendavies replied:
@dunglas do you mean `"^3.4 || ^4.0"`?

dunglas replied:
indeed

Perf commented:
Hi @dunglas,
do you know when this PR will be merged in `api-platform/core`?
I guess, you will make all `ramsey/uuid` and `api-platform/core` users happy!

@soyuka 
could you please update your fork from `api-platform/core`? It's `59 commits behind api-platform:master`.
As we don't know when this PR will be merged... at least your fork will be fully working and updated!

Thank you guys!

soyuka commented:
done

bendavies commented:
@soyuka now https://github.com/api-platform/core/pull/1781 is merged, can you think of a way to improve that, given the work you've done here?

soyuka commented:
@bendavies I'd not have merged #1781 

Instead we can fix it by adding an `IdentifierNormalizer` that supports the `__toString` method like I did [here](https://github.com/api-platform/core/pull/1478/files#diff-3c6b1ae29084aaabaf272ccf76504f0f) for ramsey uuids.

bendavies commented:
So we should use the `IdentifierNormalizer`s in the `IdentifiersExtractor`s?

soyuka commented:
Okay, I looked closer.

What I think is that we won't need your patch with this feature because the identifier can still be an Object, it'll be (de)normalized after anyway.

```
+            if (method_exists($identifiers[$propertyName], '__toString')) {
 +                $identifiers[$propertyName] = (string) $identifiers[$propertyName];
 +                continue;
 +            }
 +
```

What you do here is "denormalizing" an `UUID` normalizer to a string in a class that' responsible for caching identifiers "keys" for a given resource (because it costs a lot to find what properties are identifiers in a resource). 

What should happend with this patch:

1. Keep the object in the identifiers array
2. cast it as string with an IdentifierNormalizer after

In fact I'm going to add your test case to this patch this afternoon and see how it goes! 

> My app uses ramsey uuids for identifiers, and I noticed caching not working here. Using object identifiers effectively disabled the caching of the CachedIdentifiersExtractor, making it call the decorated IdentifiersExtractor every time.

this is something I don't get though. How is a value infering on a cache that's supposed to cache keys (identifier properties, not values) only?

bendavies commented:
I'm sorry i can't remember exactly, but `return $this->decorated->getIdentifiersFromItem($item);` was being called every time for items with ramsey uuids identifiers. my tests should reveal the problem if you revert the fix.

soyuka commented:
Updated with latest 2.2 fixes on identifier cache. We fixed it correctly so no additional changes here.

dunglas commented on composer.json:
```diff
@@ -74,6 +76,7 @@
         "guzzlehttp/guzzle": "To use the HTTP cache invalidation system.",
         "phpdocumentor/reflection-docblock": "To support extracting metadata from PHPDoc.",
         "psr/cache-implementation": "To use metadata caching.",
+        "ramsey/uuid": "To support uuid identifiers.",
```
It's not accurate, UUID are supported out of the box if generated client-side (better) or by the DBMS. I would reword this "to support Ramsey's UUID", or just drop this line.

dunglas commented on features/bootstrap/FeatureContext.php:
```diff
@@ -870,4 +871,16 @@ public function thereAreDummyImmutableDateObjectsWithDummyDate(int $nb)
+     */
+    public function thereIsARamseyIdentifiedResource(string $uuid)
+    {
+        $u = new RamseyUuidDummy();
```
`$uuid`? (explicit is better than implicit 🐍)

dunglas commented on src/Bridge/Doctrine/Orm/ItemDataProvider.php:
```diff
@@ -69,8 +70,8 @@ public function supports(string $resourceClass, string $operationName = null, ar
-
-        $identifiers = $this->normalizeIdentifiers($id, $manager, $resourceClass);
+        $hasIdentifierNormalizer = $context[IdentifierNormalizerInterface::HAS_IDENTIFIER_NORMALIZER] ?? false;
+        $identifiers = $hasIdentifierNormalizer ? $id : $this->normalizeIdentifiers($id, $manager, $resourceClass);
```
```php
if ($context[IdentifierNormalizerInterface::HAS_IDENTIFIER_NORMALIZER] ?? false) {
    $id = $this->normalizeIdentifiers($id, $manager, $resourceClass);
}
```

? (more readable)

dunglas commented on src/Bridge/Doctrine/Orm/SubresourceDataProvider.php:
```diff
@@ -152,11 +153,13 @@ private function buildQuery(array $identifiers, array $context, QueryNameGenerat
+
+        if (isset($identifiers[$identifier])) {
+            // if it's an array it's already normalized, the IdentifierManagerTrait is deprecated
+            $normalizedIdentifiers = ($context[IdentifierNormalizerInterface::HAS_IDENTIFIER_NORMALIZER] ?? false) ? $identifiers[$identifier] : $this->normalizeIdentifiers($identifiers[$identifier], $manager, $identifierResourceClass);
```
Maybe can this expression be decomposed to improve readability?

dunglas commented on src/Bridge/Ramsey/Identifier/Normalizer/UuidNormalizer.php:
```diff
@@ -0,0 +1,48 @@
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Ramsey\Identifier\Normalizer;
```
`RamseyUuid`? (in the case we integrate another Ben's library at some point)

dunglas commented on src/Bridge/Ramsey/Identifier/Normalizer/UuidNormalizer.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    public function supportsDenormalization($data, $type, $format = null)
+    {
+        return UuidInterface::class === $type || is_subclass_of($type, UuidInterface::class);
```
Maybe can you use `is_a` to remove one test (IMO it should be similar in term of perf).

soyuka replied:
`is_a` doesn't work with `is_a(Uuid::class, UuidInterface::class)` although Uuid does implement the class.

dunglas commented on src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/IdentifierNormalizerPass.php:
```diff
@@ -0,0 +1,46 @@
+    {
+        $identifierNormalizerDefinition = $container->getDefinition('api_platform.identifier_normalizer');
+
+        if (!class_exists(Uuid::class)) {
```
Should be moved in the extension.

dunglas commented on src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/IdentifierNormalizerPass.php:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @internal
+ */
+final class IdentifierNormalizerPass implements CompilerPassInterface
```
What about dropping this class for https://symfony.com/blog/new-in-symfony-3-4-simpler-injection-of-tagged-services and bumping the minimal version to 3.4?

dunglas commented on src/Bridge/Symfony/Bundle/Resources/config/api.xml:
```diff
@@ -227,6 +229,22 @@
             <argument type="service" id="api_platform.resource_class_resolver" />
         </service>
 
+        <service id="api_platform.identifier_normalizer" class="ApiPlatform\Core\Identifier\Normalizer\ChainIdentifierNormalizer" public="false">
```
`api_platform.identifier.normalizer` for consistency with the other services that have been introduced?

dunglas commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,97 @@
+
+            $type = $propertyMetadata->getType();
+            if ($type && Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType()) {
+                $metadataIdentifiers[$propertyName] = $type->getClassName();
```
`$metadataIdentifiers[$propertyName] = $type && Type::BUILTIN_TYPE_OBJECT === $type->getBuiltinType() ? $type->getClassName() : null`

dunglas commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,97 @@
+        // Normalize identifiers (DateTime, UUID etc.)
+        foreach ($identifiers as $key => $identifier) {
+            foreach ($this->identifierNormalizers as $normalizer) {
+                if ($normalizer->supportsDenormalization($identifier, $metadataIdentifiers[$key])) {
```
Maybe changing this one in a guard clause would improve the readability.

dunglas commented on src/Identifier/Normalizer/DateTimeIdentifierNormalizer.php:
```diff
@@ -0,0 +1,29 @@
+    {
+        try {
+            return parent::denormalize($data, $class, $format, $context);
+        } catch (\Exception $e) {
```
Only Symfony Serializer's `InvalidArgumentException` should be caught.

dunglas commented on src/Identifier/Normalizer/IdentifierDenormalizerInterface.php:
```diff
@@ -0,0 +1,32 @@
+use ApiPlatform\Core\Exception\InvalidIdentifierException;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+
+interface IdentifierDenormalizerInterface extends DenormalizerInterface
```
Do we really need this new interface? I would use the Symfony one directly.

soyuka replied:
Oh okay I remember. This was only to document that the `denormalize` method throws an `InvalidIdentifierException` haha. I can remove it though,

dunglas replied:
Yes please :)

dunglas commented on src/Identifier/Normalizer/IdentifierNormalizerInterface.php:
```diff
@@ -0,0 +1,21 @@
+
+namespace ApiPlatform\Core\Identifier\Normalizer;
+
+interface IdentifierNormalizerInterface
```
Why not implementing the Symfony interface here?

dunglas commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,97 @@
+        if (1 === $identifiersCount) {
+            $identifiers[$identifiersNames[0]] = $identifier;
+        } else {
+            $identifiers = $this->compositeIdentifierNormalizer->normalize($identifier);
```
To me it should be a standard normalizer, that will call recursively other ones if needed. (as we do for data normalizers)

soyuka commented on src/Bridge/Doctrine/Orm/Util/IdentifierManagerTrait.php:
```diff
@@ -39,6 +39,8 @@
      */
     private function normalizeIdentifiers($id, ObjectManager $manager, string $resourceClass): array
     {
+        @trigger_error(sprintf('Method "%s" of class "%s" is deprecated, please use identifier normalizers instead.', __METHOD__, __CLASS__), E_USER_DEPRECATED);
```
sorry @dunglas have to keep the trait it's used in 2 classes :smile:

soyuka commented:
Thanks for the review I've refactored a bit:
- uses `DenormalizeInterface` on identifier normalizers
- `CompositeIdentifierNormalizer` is now `CompositeIdentifierParser` 
- tests are green (locally at least I need to debug travis :|)
- fixed IriConverter where the flag was not set

About the deprecation message, IMO we can remove it as it's internal an no-one really uses it? We can't remove the trait because the ItemDataProvider would require `$id` to be an array, for example:
- `['id' => 1]` instead of `1`
- `['id' => 1, 'foo' => 2]` instead of `id=1;foo=2`

In fact, if we could break the DataProvider signature the best would be to type `$id` to an `array`.

dunglas commented:
> About the deprecation message, IMO we can remove it as it's internal an no-one really uses it?
👍

soyuka commented:
Any ideas for the `deps=low` test suite? I used the `tagged` thing to avoid using a pass but looks like there's something missing in `3.4`:  `Error: Class 'Symfony\Component\DependencyInjection\Argument\ClosureProxyArgument' not found`

bendavies commented:
@soyuka `symfony/event-dispatcher v3.3.0` is being installed and using  `Symfony\Component\DependencyInjection\Argument\ClosureProxyArgument` but `symfony/dependency-injection v3.4.0` is installed where it doesn't exist any more.

bendavies commented:
I think the options are, 
1. restore `symfony/dependency-injection` to `^3.3...`, restore the compiler pass and dropped the `tagged` stuff.
2. raise all low deps to `^3.4`

soyuka commented:
Green :smile:

dunglas commented:
Just great! Thanks @soyuka

redecs commented:
Thank you all for the effort put into this. It's been an issue bugging me for a while now when trying to work with UUIDs (v4).

dkarlovi commented:
\o/

meyerbaptiste commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,99 @@
+        if (\count($keys) > 1) {
+            $identifiers = $this->compositeIdentifierParser->parse($data);
+        } else {
+            $identifiers = [$keys[0] => $data];
```
What if `$keys` is empty? 💥

meyerbaptiste commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,99 @@
+        foreach ($keys as $key) {
+            foreach ($this->identifierNormalizers as $normalizer) {
+                if (!isset($identifiers[$key])) {
+                    throw new InvalidIdentifierException(sprintf('Invalid identifier "%s", "%s" was not found.', $key, $key));
```
`sprintf('... %1$s ... %1$s ...', $key)`

meyerbaptiste commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,99 @@
+    private function getIdentifierMetadata($class, $propertyName)
+    {
+        $propertyMetadata = $this->propertyMetadataFactory->create($class, $propertyName);
+        $type = $propertyMetadata->getType();
```
`$propertyMetadata` var is useless:
```php
$type = $this->propertyMetadataFactory->create($class, $propertyName)->getType();
```

meyerbaptiste commented on src/Identifier/Normalizer/ChainIdentifierNormalizer.php:
```diff
@@ -0,0 +1,99 @@
+        // Normalize every identifier (DateTime, UUID etc.)
+        foreach ($keys as $key) {
+            foreach ($this->identifierNormalizers as $normalizer) {
+                if (!isset($identifiers[$key])) {
```
Could not this condition be moved in the first loop?

meyerbaptiste commented on src/Identifier/Normalizer/DateTimeIdentifierNormalizer.php:
```diff
@@ -0,0 +1,31 @@
+        try {
+            return parent::denormalize($data, $class, $format, $context);
+        } catch (InvalidArgumentException $e) {
+            throw new InvalidIdentifierException($e->getMessage());
```
You should populate the previous exception argument.

meyerbaptiste commented on .travis.yml:
```diff
@@ -41,7 +41,7 @@ before_install:
   - echo "memory_limit=-1" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
   - npm install -g swagger-cli
-  - if [[ $lint = 1 ]]; then wget https://github.com/FriendsOfPHP/PHP-CS-Fixer/releases/download/v2.8.2/php-cs-fixer.phar; fi
+  - if [[ $lint = 1 ]]; then wget https://github.com/FriendsOfPHP/PHP-CS-Fixer/releases/download/v2.8.4/php-cs-fixer.phar; fi
```
I think 2.10.0 should work!

soyuka commented on src/Bridge/Symfony/Routing/IriConverter.php:
```diff
@@ -43,12 +45,14 @@
         $this->itemDataProvider = $itemDataProvider;
         $this->routeNameResolver = $routeNameResolver;
         $this->router = $router;
+        $this->identifierNormalizer = $identifierNormalizer;
```
nope we check below we don't want to break this class.

cve commented:
#1478 is not it better?
